tests/benchmarks/data_structure_perf/rbtree_perf/src/rbtree_perf.c

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright (c) 2017 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <ztest.h>
+#include <sys/rb.h>
+
+#define TREE_SIZE 512
+
+/* rbnode structure is embeddedable in user structure */
+struct container_node {
+	struct rbnode node;
+	int value;
+};
+static struct rbnode nodes[TREE_SIZE];
+static struct rbtree tree;
+
+/* Our "lessthan" is just the location of the struct */
+bool node_lessthan(struct rbnode *a, struct rbnode *b)
+{
+	return a < b;
+}
+
+/**
+ * @brief Test whether rbtree node struct is embedded
+ * in any user struct
+ *
+ * @details Initialize an user defined structure contains
+ * rbtree node.And into a rbtree and enumerate
+ * the rbtree nodes.
+ *
+ * verify that the value enumerated is equal to the
+ * value initialized.If the verification pass,the user
+ * defined structure works.
+ * verify "for each" style APIs work.
+ *
+ * @ingroup lib_rbtree_tests
+ *
+ * @see RB_FOR_EACH(),RB_FOR_EACH_CONTAINER()
+ */
+void test_rbtree_container(void)
+{
+	int count = 0;
+	static struct rbtree test_tree_l;
+	static struct container_node tree_node[10];
+	static struct container_node *c_foreach_node;
+	static struct rbnode *foreach_node;
+
+	test_tree_l.lessthan_fn = node_lessthan;
+	for (unsigned int i = 0; i < ARRAY_SIZE(tree_node); i++) {
+		tree_node[i].value = i;
+		rb_insert(&test_tree_l, &tree_node[i].node);
+	}
+
+	RB_FOR_EACH(&test_tree_l, foreach_node) {
+		zassert_true(CONTAINER_OF(foreach_node, struct container_node,
+		node)->value == count, "RB_FOR_EACH failed");
+		count++;
+	}
+
+	count = 0;
+
+	RB_FOR_EACH_CONTAINER(&test_tree_l, c_foreach_node, node) {
+		zassert_true(c_foreach_node->value == count,
+		"RB_FOR_EACH_CONTAINER failed");
+		count++;
+	}
+}
+
+/* initalize and insert a tree */
+void init_tree(struct rbtree *tree, int size)
+{
+	tree->lessthan_fn = node_lessthan;
+
+	for (int i = 0; i < size; i++) {
+		rb_insert(tree, &nodes[i]);
+	}
+}
+
+uint32_t node_height;
+int search_height_recurse(struct rbnode *node, struct rbnode *finial_node)
```

Comment:
I would change this algorithm, you don't need to rely on a global variable.
Just add this parameter to the function to keep the height track, something like:
```
int search_height_recurse(struct rbnode *node, struct rbnode *final_node, uint32_t current_height)
{
	if (node == NULL) {
		return -1;
	}

	if (node == final_node) {
		return current_height;
 	}

	current_height++;
	struct rbnode *ch = z_rb_child(node, tree.lessthan_fn(finial_node, node));
	return search_height_recurse(ch, final_node, current_height);
}
```

Note, I haven't check the algorithm above accuracy, just the idea. Then you don't need to keep that global, you just call this function with initial value 0.

Reply:
The code looks great, and I do a few modifications to make it can work better. I have used the code,thanks.